### ISSUE TITLE
Check the well-formed-ness of type aliases.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#44075]
Currently, these compile:
```rust
struct Foo<T: Copy>(T);
type X = Foo<String>;
type Y = <() as Iterator>::Item;
```

Using them does produce an error *at the use site*, but not in the definition.

**Comment by eddyb** [comment]
If we do check WF for type aliases, can we also easily check if there are bounds *not used* to satisfy well-formed-ness rules? Because those are the bounds we current ignore where the type alias is used.

**Comment by nikomatsakis** [comment]
@eddyb not super easily

**Comment by eddyb** [comment]
For the record, *it's actually easy*. This is what it looks like on the current compiler:
```rust
let user_predicates = fcx.param_env.caller_bounds;
let wf_predicates = ty::wf::obligations(
    &fcx.infcx,
    fcx.param_env,
    fcx.body_id,
    item_ty,
    item.span,
).into_iter().flatten().map(|o| o.predicate).collect();
let wf_predicates = traits::elaborate_predicates(fcx.tcx, wf_predicates);
let wf_param_env = ty::ParamEnv {
    caller_bounds: fcx.tcx.mk_predicates(wf_predicates),
    reveal: fcx.param_env.reveal,
};
fcx.register_predicates(user_predicates.iter().map(|&predicate| {
    let code = ObligationCauseCode::MiscObligation;
    let cause = traits::ObligationCause::new(item.span, fcx.body_id, code);
    traits::Obligation::new(cause, wf_param_env, predicate)
}));
```

Of course, proper reporting is more work, but it's a shame we didn't figure out we can do these checks, in the first place, last year, instead of so close to the edition release (https://github.com/rust-lang/rust/issues/49441#issuecomment-414470068).
